src/levenshtein.cr

```diff
@@ -13,36 +13,70 @@ module Levenshtein
   def self.distance(string1 : String, string2 : String) : Int32
     return 0 if string1 == string2
 
-    s = string1.chars
-    t = string2.chars
-
-    s_size = s.size
-    t_size = t.size
+    s_size = string1.size
+    t_size = string2.size
 
     return t_size if s_size == 0
     return s_size if t_size == 0
 
-    # This is to allocate less memory
-    if t_size > s_size
-      t, s = s, t
-      t_size, s_size = s_size, t_size
-    end
+    if string1.ascii_only? && string2.ascii_only?
+      s = string1.to_unsafe
+      t = string2.to_unsafe
+
+      # This is to allocate less memory
+      if t_size > s_size
+        t, s = s, t
+        t_size, s_size = s_size, t_size
+      end
 
-    v = Pointer(Int32).malloc(t_size + 1) { |i| i }
+      v = Pointer(Int32).malloc(t_size + 1) { |i| i }
 
-    s_size.times do |i|
-      last_cost = i + 1
+      i = 0
+      last_cost = 0
+      s_size.times do |i|
+        last_cost = i + 1
 
-      t_size.times do |j|
-        sub_cost = s[i] == t[j] ? 0 : 1
-        cost = Math.min(Math.min(last_cost + 1, v[j + 1] + 1), v[j] + sub_cost)
-        v[j] = last_cost
-        last_cost = cost
+        t_size.times do |j|
+          sub_cost = s[i] == t[j] ? 0 : 1
+          cost = Math.min(Math.min(last_cost + 1, v[j + 1] + 1), v[j] + sub_cost)
+          v[j] = last_cost
+          last_cost = cost
+        end
+        v[t_size] = last_cost
       end
-      v[t_size] = last_cost
-    end
 
-    v[t_size]
+      last_cost
+    else
+      reader1 = Char::Reader.new(string1)
+      reader2 = Char::Reader.new(string2)
```

Comment:
Could you provide a benchmark for large non-ascii strings with `chars` vs. `Char::Reader`? With `Char::Reader` it means we are decoding the second string many times, whereas before it was just decoded once, and turns into an array of chars.

Reply:
```cr
require "benchmark"

Benchmark.ips do |bm|
  bm.report "1" do
    Levenshtein.distance1("🐓" * 100000, "hello")
    Levenshtein.distance1("hello", "🐓" * 100000)
  end

  bm.report "2" do
    Levenshtein.distance2("🐓" * 100000, "hello")
    Levenshtein.distance2("hello", "🐓" * 100000)
  end
end
```
```
      .chars 253.65  (  3.94ms) (± 8.42%)  1.53MB/op        fastest
Char::Reader 184.69  (  5.41ms) (± 7.57%)   781kB/op   1.37× slower
```
Yes for large non-ASCII strings `.chars` is a bit faster.
Even for small non-ASCII strings (25 instead of 100000):
```
      .chars 682.39k (  1.47µs) (± 8.06%)  736B/op        fastest
Char::Reader 592.25k (  1.69µs) (± 9.69%)  320B/op   1.15× slower
```
But the important difference is also the memory usage. `.chars` uses a lot more memory than the `Char::Reader` way.

But actually, what if I *combine* `.chars` and `Char::Reader` and use `.chars` only for the second string?

```cr
require "benchmark"

Benchmark.ips do |bm|
  bm.report "2x .chars" do
    Levenshtein.distance1("🐓" * 100000, "hello")
    Levenshtein.distance1("hello", "🐓" * 100000)
  end

  bm.report "1x Char::Reader + 1x .chars" do
    Levenshtein.distance2("🐓" * 100000, "hello")
    Levenshtein.distance2("hello", "🐓" * 100000)
  end
end
```
```
                  2x .chars 522.71  (  1.91ms) (± 9.40%)  782kB/op   1.42× slower
1x Char::Reader + 1x .chars 740.84  (  1.35ms) (± 9.90%)  391kB/op        fastest
```
It seems like a good balance between memory usage and speed.
25 instead of 100000:
```
                  2x .chars 555.75k (  1.80µs) (±10.41%)  736B/op   1.50× slower
1x Char::Reader + 1x .chars 835.18k (  1.20µs) (±12.23%)  528B/op        fastest
```
What do you think?

Reply:
Yes, mixing strategy would be great!